include/rsp_queue.inc

```diff
@@ -0,0 +1,566 @@
+    li a0, RSPQ_HIGHPRI_CALL_SLOT<<2
+    li a1, RSPQ_LOWPRI_CALL_SLOT<<2
+    sub rspq_dmem_buf_ptr, t0
+    #fallthrough
+    .endfunc
+
+    #############################################################
+    # RSPQCmd_SwapBuffers
+    #
+    # Switch between lowpri and highpri or viceversa. This is
+    # called by RSP itself to go into highpri mode, and scheduled
+    # as normal command by CPU when going back into lowpri.
+    #
+    # ARGS:
+    #   a0: Pointer stack slot that contains the address to switch to.
+    #       (either RSPQ_LOWPRI_CALL_SLOT<<2 or RSPQ_HIGHPRI_CALL_SLOT<<2)
+    #   a1: Pointer stack slot where to save the current address to.
+    #       (either RSPQ_LOWPRI_CALL_SLOT<<2 or RSPQ_HIGHPRI_CALL_SLOT<<2)
+    #   a2: New mask to check for HIGHPRI signal (0 in highpri mode).
+    #############################################################
+    .func RSPQCmd_SwapBuffers
+RSPQCmd_SwapBuffers:
+    mtc0 a2, COP0_SP_STATUS
+    lw a0, %lo(RSPQ_POINTER_STACK)(a0)
+    #fallthrough
+    .endfunc    
+    
+    #############################################################
+    # RSPQCmd_Call
+    #
+    # Do a nested call to a different command list. Save the
+    # current RDRAM position to be able to resume execution
+    # later.
+    #
+    # ARGS:
+    #   a0: New RDRAM address (plus command opcode)
+    #   a1: DMEM address of the save slot for the current address
+    #############################################################
+    .func RSPQCmd_Call
+RSPQCmd_Call:
+    # a0: command opcode + RDRAM address
+    # a1: call slot in DMEM
+    lw s0, %lo(RSPQ_RDRAM_PTR)
+    add s0, rspq_dmem_buf_ptr
+    sw s0, %lo(RSPQ_POINTER_STACK)(a1)  # save return address
+    # fallthrough
+    .endfunc
+
+    #############################################################
+    # RSPQCmd_Jump
+    #
+    # Jump to a different command list. The current RDRAM position
+    # is lost.
+    #
+    # ARGS:
+    #   a0: New RDRAM address (plus command opcode)
+    #############################################################
+    .func RSPQCmd_Jump
+RSPQCmd_Jump:
+    # a0: command opcode + RDRAM address
+    j rspq_fetch_buffer_with_ptr
+    move s0, a0
+    .endfunc
+
+    #############################################################
+    # RSPQCmd_Ret
+    #
+    # Restore a previously saved RDRAM position and jump to it.
+    #
+    # ARGS:
+    #   a0: DMEM address of the save slot (plus command opcode)
+    #############################################################
+    .func RSPQCmd_Ret
+RSPQCmd_Ret:
+    # a0: command opcode + call slot in DMEM to recover
+    j rspq_fetch_buffer_with_ptr
+    lw s0, %lo(RSPQ_POINTER_STACK)(a0)
+    .endfunc
+
+    #############################################################
+    # RSPQCmd_TestWriteStatus
+    #
+    # Like RSPQCmd_WriteStatus, writes COP0 SP status register with a specified value,
+    # but first waits until the SP status AND'ed with the specified bitmask is zero.
```

Comment:
Yes, reading and writing the SP status register is atomic. The only instance where that matters are the signals though I think, which can be set/unset arbitrarily by both CPU and RSP. This command is primarily used for synpoints. In addition to raising an interrupt, syncpoints will set a signal when triggered (`SP_STATUS_SIG_SYNCPOINT`). Before triggering a syncpoint, this command will check if that signal is still set (it gets reset by CPU when handling the interrupt). This is done to prevent these interrupts from being coalesced, which would cause syncpoints to be skipped.

I'm not sure which RDRAM writes you are referring to? I think the SP status register works differently though, since it is wired up directly to a specific virtual address instead of going through RDRAM.

Reply:
It is possible to dispatch 32bit memory mapped writes to the RDRAM which all go through RI and I thought there might be something in common between CPU-RI and CPU-SP_STATUS that does the arbitration, probably this is a different mechanism though. Maybe @rasky knows.
Regardless, the important thing these being atomic, which means either SPU or RSP stalls when the other is dispatching an SP_STATUS update. I wonder how this works :)
Probably a non-issue for this PR if we know it is atomic.

Reply:
I'm not sure what you're referring to, wrt RDRAM.

Writes to SP_STATUS are atomic as in the fact that the write masks have separate set/reset flags means that you don't need a read-modify-write cycle which can introduce race conditions. For instance, let's assume that SP_STATUS didn't even the special SP_WSTATUS_* bits and could just be written like a regular memory location, using SP_STATUS_* bits. To turn off SP_STATUS_SIG_SYNCPOINT,  the CPU would have to do:

```C
   *SP_STATUS = *SP_STATUS & ~SP_STATUS_SIG_SYNCPOINT;
```

which means that the CPU would first read the register and then write it. If the RSP races with this and turn on a different bit, that write would be lost.

Instead, given that the CPU can simply do `*SP_STATUS = SP_WSTATUS_CLEAR_SIG_SYNCPOINT` and everything is handled internally, it means that the RSP can at the same time turn on a different bit and everything works as expected. Likely, what happens if the two memory writes happens exactly at the same clock cycle, is that either of the two gets stalled one cycle more, and so they are actually serialized.

Obviously, this doesn't mean that modifying *the same bit* by both CPU and RSP has a well defined behavior. That is a true race condition where there is no amount of hardware suppor that can save you. If the CPU is setting one bit and the RSP is resetting it at the same time, it's undefined behavior whether the bit ends up being set or cleared. After all, there is no way to synchronize the two cores, so what we think it's happening "at the same time" could be slightly drifted in any direction, and we would have no way to prove it or control it. 

In other words, this is the same guarantee of an atomic variable in modern programming. You can surely access it from two threads without any other locking, but you must behave well while accessing it, following a programming pattern that takes into account that either side can access it at any moment.

Reply:
wrt RDRAM what I meant was it is not possible to partially write a 32 bit value by CPU memory-mapped writes.

> Writes to SP_STATUS are atomic as in the fact that the write masks have separate set/reset flags means that you don't need a read-modify-write cycle which can introduce race conditions.

This is totally true if different cores were accessing on-chip SP_STATUS interfaces. But in this case the RSP is a separate chip and when the CPU tries to set some bits via the SP_STATUS flags, they need to propagate to the RSP chip and will actually physically change bits on some digital interface. If this digital interface is physically the same with the one used internally by RSP (it probably is) then there need to be a mechanism to tell the CPU that this is being written to stall it or an intermediate buffer to keep the flags coming from the CPU until the RSP operation finishes, if any.

If the CPU-RSP interface is not 32 bit and instead if the CPU is setting the flags 8 bits at a time and there isn't a mechanism to sync the two, RSP can partially override the *flags* that are being written before they are latched to update the register state.

The same is probably not a concern when reading the register contents even if the CPU can execute the flags partially because we are not interested in flags changing in tandem. Though it would mean halt and sig more may not be set together causing other problems...

But if the propagation of the flags from CPU to RSP is in 32bit chunks (like how RI interface works - 32bit at a time - like we are certain we cannot write a partial command to the RDRAM) while mutually stalling each other we are all fine! This is what I was trying to explain with a few poorly written sentences before 😆 

 >  everything is handled internally, it means that the RSP can at the same time turn on a different bit and everything works as expected

Then we are allright! I know I maybe overthinking a little but this is N64 with its various hardware quirks!

> If the CPU is setting one bit and the RSP is resetting it at the same time, it's undefined behavior

This made me think of a potential scenario;

- CPU kicks off the first sig more
- RSP loads this update by `mfc0 t0, COP0_SP_STATUS` and decides to wake up
- CPU kicks off second sig more in tandem with `mtc0 t0, COP0_SP_STATUS`

If RSP wins, sig more is reset and RSP will sleep when it is done.

If CPU wins, sig more is set, even though the RSP is already awake, potentially trying to fetch new commands and this may cause partially written commands to execute?

I really hope I am wrong on this one.